### PR TITLE
Implementation of HD2017 dust model

### PR DESCRIPTION
**Pull request by zonca**

### CODE REVIEW

**Comment by zonca** [comment]
@bthorne93 I'm finally back working on this, I have implemented unit tests, they don't pass!

* I created PySM 2 HD2017 maps with https://gist.github.com/zonca/a1e35bce9d467519e94666524daf0df2#file-hd17_pysm2-ipynb
* saved FITS files to https://github.com/healpy/pysm-data/tree/master/pysm_2_test_data
* compared PySM 2 and 3, in `uK_RJ` and they have a disagreement up to 10%, see https://gist.github.com/zonca/a1e35bce9d467519e94666524daf0df2#file-hd17_pysm2-pysm3-ipynb

PySM 3 does the calculation on float32 to save memory, but I don't think that is enough to explain the difference, what do you think?

**Comment by bthorne93** [comment]
@zonca My first guess would be that this is just a result of the random generation of part of the spectral model using a different seed. We would need to check exactly how `check_d7_353_64.fits` was produced. Was this in the original pysm2 testing directory? (I can't remember if we'd implemented it at that point)

**Comment by zonca** [comment]
I created check_d7_353_64.fits
And used the same seed, is it possible the seed has different assumptions?

**Comment by zonca** [comment]
See the notebook above here about how I created that test file

File: pysm/data/presets.cfg [hunk 1]
```diff
@@ -90,6 +90,20 @@ map_mbb_index = "pysm_2/dust_beta.fits"
...
+map_I = "pysm_2/dust_t_new.fits"
+map_Q = "pysm_2/dust_q_new.fits"
+map_U = "pysm_2/dust_u_new.fits"
+unit_I = "uK_RJ"
+unit_Q = "uK_RJ"
+unit_U = "uK_RJ"
+freq_ref_I = "545 GHz"
+freq_ref_P = "353 GHz"
+rnd_uval = true
+seed = 4632
```
Comment (zonca) [inline pysm/data/presets.cfg:104]: @bthorne93 
Compare to PySM2: https://github.com/bthorne93/PySM_public/blob/master/pysm/nominal.py#L122

**Comment by bthorne93** [comment]
Okay, doesn't seem to be that, the code setting the seed and doing the random generating is identical in the two versions. I'll take a deeper look now.

File: pysm/models/dust.py [hunk 2]
```diff
@@ -366,3 +368,393 @@ def blackbody_nu(freq, temp):
...
+        assert car_data_p.unit == u.Jy / u.sr * u.cm ** 2
+        self.car_p = RectBivariateSpline(uvec, wav, car_data_p.T)
+
+        assert silfe_data_p.unit == u.Jy / u.sr * u.cm ** 2
+        self.silfe_p = RectBivariateSpline(uvec, wav, silfe_data_p.T)
+
+        # now draw the random realisation of uval if draw_uval = true
+        if rnd_uval:
+            T_mean = self.read_map(
+                "pysm_2/COM_CompMap_dust-commander_0256_R2.00.fits", unit="K", field=3, nside=256
```
[the remark this answers is not in the document]
Comment (bthorne93) [reply pysm/models/dust.py:554]: @zonca that sounds reasonable, will do

[the remark this answers is not in the document]
Comment (bthorne93) [reply pysm/models/dust.py]: Done

[the remark this answers is not in the document]
Comment (zonca) [reply pysm/models/dust.py]: thanks! can you add to docstring as well?

**Comment by zonca** [comment]
@bthorne93 can you also rebase on master, so the unit tests run

File: pysm/models/dust.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bthorne93) [reply pysm/models/dust.py]: Sorry, bad habits die hard.

**Comment by bthorne93** [comment]
@zonca Since it's a destructive operation, and I'm not comfortable enough with my git skills, would you mind doing that? I've added the docstring above.

**Comment by zonca** [comment]
@bthorne93 sure, no problem

**Comment by zonca** [comment]
@bthorne93 rebase was too messy, I merged!

tests pass on my machine!

**Comment by bthorne93** [comment]
Great!

**Comment by zonca** [comment]
no, I'll make sure everything passes on travis then merge

**Comment by zonca** [comment]
Not sure why github is not showing the travis builds feedback anymore.
Anyway, all tests pass:

https://travis-ci.org/github/healpy/pysm/builds/670830673

excellent, thanks @bthorne93 !

**Pull request merged**